File: components/MAVLink/MAVLink.c
#include <stdio.h>
#include "MAVLink.h"

// CRC accumulation function (X.25 standard)
void crc_accumulate(uint8_t data, uint16_t* crcAccum) {
    uint8_t tmp = data ^ (*crcAccum & 0xFF);
    tmp ^= (tmp << 4);
    *crcAccum = (*crcAccum >> 8) ^ ((uint16_t)tmp << 8) ^ ((uint16_t)tmp << 3) ^ ((uint16_t)tmp >> 4);
}

// Initialize CRC
void crc_init(uint16_t* crcAccum) {
    *crcAccum = X25_INIT_CRC;
}

// Calculate CRC over a buffer
uint16_t crc_calculate(const uint8_t* pBuffer, uint16_t length) {
    uint16_t crcTmp;
    crc_init(&crcTmp);
    while (length--) {
        crc_accumulate(*pBuffer++, &crcTmp);
    }
    return crcTmp;
}

// Accumulate CRC over a buffer
void crc_accumulate_buffer(uint16_t* crcAccum, const uint8_t* pBuffer, uint16_t length) {
    while (length--) {
        crc_accumulate(*pBuffer++, crcAccum);
    }
}

// Start checksum calculation for a message
void mavlink_start_checksum(mavlink_message_t* msg) {
    crc_init(&msg->checksum);
}

// Update checksum with a single byte
void mavlink_update_checksum(mavlink_message_t* msg, uint8_t c) {
    crc_accumulate(c, &msg->checksum);
}

// Encode a MAVLink message into a buffer
int mavlink_encode(const mavlink_message_t* msg, uint8_t* buffer) {
    int offset = 0;

    // Header
    buffer[offset++] = MAVLINK_STX;      // Start byte
    buffer[offset++] = msg->len;         // Payload length
    buffer[offset++] = msg->seq;         // Sequence number
    buffer[offset++] = msg->sysid;       // System ID
    buffer[offset++] = msg->compid;      // Component ID
    buffer[offset++] = msg->msgid;       // Message ID
    // buffer[offset++] = msg->segment_index; // Segment index
    // buffer[offset++] = msg->total_segments; // Total segments

    // Initialize checksum (after STX)
    mavlink_start_checksum((mavlink_message_t*)msg);
    mavlink_update_checksum((mavlink_message_t*)msg, msg->len);
    mavlink_update_checksum((mavlink_message_t*)msg, msg->seq);
    mavlink_update_checksum((mavlink_message_t*)msg, msg->sysid);
    mavlink_update_checksum((mavlink_message_t*)msg, msg->compid);
    mavlink_update_checksum((mavlink_message_t*)msg, msg->msgid);
    // mavlink_update_checksum((mavlink_message_t*)msg, msg->segment_index);
    // mavlink_update_checksum((mavlink_message_t*)msg, msg->total_segments);

    // Payload
    for (int i = 0; i < msg->len; ++i) {
        buffer[offset++] = msg->payload[i];
        mavlink_update_checksum((mavlink_message_t*)msg, msg->payload[i]);
    }

    // Add CRC extra (placeholder; define your own table or remove if not needed)
#ifdef MAVLINK_CRC_EXTRA
    // Example: Replace with actual message-specific CRC extra from mavlink library
    uint8_t crc_extra = 0; // Placeholder (e.g., MAVLINK_MESSAGE_CRC_(msg->msgid))
    mavlink_update_checksum((mavlink_message_t*)msg, crc_extra);
#endif

    // Append checksum
    buffer[offset++] = msg->checksum & 0xFF;        // Low byte
    buffer[offset++] = (msg->checksum >> 8) & 0xFF; // High byte

    return offset; // Total length of encoded message
}

// Decode a MAVLink message from a buffer
int mavlink_decode(mavlink_message_t* rxmsg, const uint8_t* buffer, uint16_t length) {
    mavlink_state_t state = MAVLINK_STATE_IDLE;
    int pos = 0;

    if (length < 8) { // Minimum: STX + len + seq + sysid + compid + msgid + CRC (2 bytes)
        return -1; // Buffer too short
    }

    while (pos < length && state != MAVLINK_STATE_GOT_CRC1) {
        switch (state) {
            case MAVLINK_STATE_IDLE:
                if (buffer[pos] == MAVLINK_STX) {
                    rxmsg->header = buffer[pos];
                    mavlink_start_checksum(rxmsg);
                    state = MAVLINK_STATE_GOT_STX;
                    pos++;
                } else {
                    return -1; // Invalid start byte
                }
                break;

            case MAVLINK_STATE_GOT_STX:
                rxmsg->len = buffer[pos];
                mavlink_update_checksum(rxmsg, buffer[pos]);
                state = MAVLINK_STATE_GOT_LENGTH;
                pos++;
                break;

            case MAVLINK_STATE_GOT_LENGTH:
                rxmsg->seq = buffer[pos];
                mavlink_update_checksum(rxmsg, buffer[pos]);
                state = MAVLINK_STATE_GOT_SEQ;
                pos++;
                break;

            case MAVLINK_STATE_GOT_SEQ:
                rxmsg->sysid = buffer[pos];
                mavlink_update_checksum(rxmsg, buffer[pos]);
                state = MAVLINK_STATE_GOT_SYSID;
                pos++;
                break;

            case MAVLINK_STATE_GOT_SYSID:
                rxmsg->compid = buffer[pos];
                mavlink_update_checksum(rxmsg, buffer[pos]);
                state = MAVLINK_STATE_GOT_COMPID;
                pos++;
                break;

            case MAVLINK_STATE_GOT_COMPID:
                rxmsg->msgid = buffer[pos];
                mavlink_update_checksum(rxmsg, buffer[pos]);
                state = MAVLINK_STATE_GOT_MSGID;
                pos++;
                break;

            case MAVLINK_STATE_GOT_MSGID:
                if (pos + rxmsg->len + 2 > length) {
                    return -1; // Buffer too short for payload + CRC
                }
                for (int i = 0; i < rxmsg->len; ++i) {
                    rxmsg->payload[i] = buffer[pos + i];
                    mavlink_update_checksum(rxmsg, buffer[pos + i]);
                }
#ifdef MAVLINK_CRC_EXTRA
                // Placeholder: Replace with actual message-specific CRC extra
                uint8_t crc_extra = 0; // e.g., MAVLINK_MESSAGE_CRC_(rxmsg->msgid)
                mavlink_update_checksum(rxmsg, crc_extra);
#endif
                pos += rxmsg->len;
                state = MAVLINK_STATE_GOT_PAYLOAD;
                break;

            // case MAVLINK_STATE_GOT_SEGMENT_INDEX:
            //     rxmsg->msgid = buffer[pos];
            //     mavlink_update_checksum(rxmsg, buffer[pos]);
            //     state = MAVLINK_STATE_GOT_MSGID;
            //     pos++;
            //     break;

            // case MAVLINK_STATE_GOT_TOTAL_SEGMENTS:
            //     rxmsg->msgid = buffer[pos];
            //     mavlink_update_checksum(rxmsg, buffer[pos]);
            //     state = MAVLINK_STATE_GOT_MSGID;
            //     pos++;
            //     break;

            case MAVLINK_STATE_GOT_PAYLOAD:
                rxmsg->checksum_calculator[0] = buffer[pos];
                rxmsg->checksum_calculator[1] = buffer[pos + 1];
                state = MAVLINK_STATE_GOT_CRC1;
                pos += 2;
                break;

            default:
                return -1; // Invalid state
        }
    }

    if (state != MAVLINK_STATE_GOT_CRC1) {
        return -1; // Incomplete message
    }

    // Verify checksum
    uint8_t crc_temp[2];
    crc_temp[0] = rxmsg->checksum & 0xFF;       // Low byte
    crc_temp[1] = (rxmsg->checksum >> 8) & 0xFF; // High byte
    if (crc_temp[0] == rxmsg->checksum_calculator[0] && crc_temp[1] == rxmsg->checksum_calculator[1]) {
        return pos; // Success, return number of bytes processed
    } else {
        return -1; // Checksum mismatch
    }
}

File: python_client_interface/Lib/MAVLink.py
# Các hằng số cần thiết
X25_INIT_CRC = 0xFFFF
MAVLINK_CRC_EXTRA = False  # Nếu có dùng CRC extra, đổi thành True

# Ví dụ về hàm trả về CRC extra cho msgid nào đó.
# Trong thực tế, hàm này sẽ trả về một giá trị phụ thuộc vào msgid.
def MAVLINK_MESSAGE_CRC_(msgid):
    # Ví dụ: trả về một giá trị cố định hay lookup từ bảng
    return 0  # Thay đổi nếu cần


# Hàm tính toán CRC cho từng byte (tương đương crc_accumulate)
def crc_accumulate(data, crcAccum):
    # data: một số nguyên 0<=data<256
    # crcAccum: số nguyên 0<=crcAccum<65536
    tmp = data ^ (crcAccum & 0xFF)
    tmp = tmp ^ ((tmp << 4) & 0xFF)
    new_crc = (crcAccum >> 8) ^ ((tmp & 0xFF) << 8) ^ ((tmp & 0xFF) << 3) ^ ((tmp & 0xFF) >> 4)
    return new_crc & 0xFFFF


# Khởi tạo crc
def crc_init():
    return X25_INIT_CRC


# Tính toán CRC cho một buffer (danh sách hoặc bytes)
def crc_calculate(pBuffer):
    crcTmp = crc_init()
    for b in pBuffer:
        crcTmp = crc_accumulate(b, crcTmp)
    return crcTmp


# Tích lũy CRC cho một buffer (buffer truyền vào kiểu chuỗi hoặc danh sách các ký tự)
def crc_accumulate_buffer(crcAccum, pBuffer):
    # pBuffer: chuỗi hoặc danh sách ký tự
    for ch in pBuffer:
        # Nếu pBuffer là chuỗi, chuyển ký tự sang mã ascii (0<=x<256)
        if isinstance(ch, str):
            b = ord(ch)
        else:
            b = ch
        crcAccum = crc_accumulate(b, crcAccum)
    return crcAccum


# Định nghĩa lớp đại diện cho thông điệp MAVLink
class MavlinkMessage:
    def __init__(self):
        self.header = 0         # Start byte
        self.len = 0            # Độ dài payload
        self.seq = 0            # Số thứ tự
        self.sysid = 0          # System ID
        self.compid = 0         # Component ID
        self.msgid = 0          # Message ID
        self.payload = []       # Danh sách các byte (số nguyên 0<=x<256)
        self.checksum = 0       # Giá trị checksum (tính toán trong quá trình encode/decode)
        self.checksum_calculator = [0, 0]  # Checksum nhận được (2 byte)


# Các hàm hỗ trợ checksum cho MAVLink
def mavlink_start_checksum(msg):
    msg.checksum = crc_init()


def mavlink_update_checksum(msg, c):
    msg.checksum = crc_accumulate(c, msg.checksum)


# Các trạng thái của máy trạng thái khi decode
MAVLINK_STATE_IDLE      = 0
MAVLINK_STATE_GOT_STX   = 1
MAVLINK_STATE_GOT_LENGTH = 2
MAVLINK_STATE_GOT_SEQ   = 3
MAVLINK_STATE_GOT_SYSID = 4
MAVLINK_STATE_GOT_COMPID = 5
MAVLINK_STATE_GOT_MSGID = 6
MAVLINK_STATE_GOT_PAYLOAD = 7
MAVLINK_STATE_GOT_CRC   = 8
MAVLINK_STATE_GOT_CRC1  = 9


def mavlink_decode(rxmsg, c):
    """
    Giải mã một thông điệp MAVLink từ danh sách các byte (c).
    rxmsg: đối tượng MavlinkMessage đã được khởi tạo
    c: danh sách các byte (số nguyên 0<=x<256) chứa thông điệp đầy đủ
       dự kiến định dạng: [header, len, seq, sysid, compid, msgid, payload[0...len-1], crc_low, crc_high]
    Trả về True nếu checksum đúng, False nếu sai.
    """
    state = MAVLINK_STATE_IDLE

    # Dùng chỉ số truy cập vào danh sách c (giá trị cố định theo định dạng MAVLink)
    while state != MAVLINK_STATE_GOT_CRC1:
        if state == MAVLINK_STATE_IDLE:
            mavlink_start_checksum(rxmsg)
            state = MAVLINK_STATE_GOT_STX

        elif state == MAVLINK_STATE_GOT_STX:
            rxmsg.header = c[0]
            state = MAVLINK_STATE_GOT_LENGTH

        elif state == MAVLINK_STATE_GOT_LENGTH:
            rxmsg.len = c[1]
            mavlink_update_checksum(rxmsg, c[1])
            state = MAVLINK_STATE_GOT_SEQ

        elif state == MAVLINK_STATE_GOT_SEQ:
            rxmsg.seq = c[2]
            mavlink_update_checksum(rxmsg, c[2])
            state = MAVLINK_STATE_GOT_SYSID

        elif state == MAVLINK_STATE_GOT_SYSID:
            rxmsg.sysid = c[3]
            mavlink_update_checksum(rxmsg, c[3])
            state = MAVLINK_STATE_GOT_COMPID

        elif state == MAVLINK_STATE_GOT_COMPID:
            rxmsg.compid = c[4]
            mavlink_update_checksum(rxmsg, c[4])
            state = MAVLINK_STATE_GOT_MSGID

        elif state == MAVLINK_STATE_GOT_MSGID:
            rxmsg.msgid = c[5]
            mavlink_update_checksum(rxmsg, c[5])
            state = MAVLINK_STATE_GOT_PAYLOAD

        elif state == MAVLINK_STATE_GOT_PAYLOAD:
            rxmsg.payload = []
            for i in range(rxmsg.len):
                byte = c[6 + i]
                rxmsg.payload.append(byte)
                mavlink_update_checksum(rxmsg, byte)
            if MAVLINK_CRC_EXTRA:
                mavlink_update_checksum(rxmsg, MAVLINK_MESSAGE_CRC_(rxmsg.msgid))
            state = MAVLINK_STATE_GOT_CRC

        elif state == MAVLINK_STATE_GOT_CRC:
            rxmsg.checksum_calculator = [ c[6 + rxmsg.len], c[7 + rxmsg.len] ]
            state = MAVLINK_STATE_GOT_CRC1

        else:
            # Nếu có trạng thái không xác định, thoát vòng lặp
            break

    # Tách checksum tính được thành 2 byte
    crc_low  = rxmsg.checksum & 0xFF
    crc_high = (rxmsg.checksum >> 8) & 0xFF

    if (crc_low == rxmsg.checksum_calculator[0] and
        crc_high == rxmsg.checksum_calculator[1]):
        return True  # Thành công
    else:
        return False  # Checksum không khớp


def mavlink_encode(msg, buffer):
    """
    Mã hoá thông điệp MAVLink từ đối tượng msg (MavlinkMessage)
    và ghi kết quả vào buffer (danh sách, được mở rộng theo nhu cầu).
    Trả về chiều dài của thông điệp đã mã hoá.
    """
    offset = 0

    # Ghi header và các trường khác
    buffer.append(msg.header)   ; offset += 1
    buffer.append(msg.len)      ; offset += 1
    buffer.append(msg.seq)      ; offset += 1
    buffer.append(msg.sysid)    ; offset += 1
    buffer.append(msg.compid)   ; offset += 1
    buffer.append(msg.msgid)    ; offset += 1

    # Khởi tạo checksum
    mavlink_start_checksum(msg)

    # Tích lũy checksum cho header và message ID
    mavlink_update_checksum(msg, msg.len)
    mavlink_update_checksum(msg, msg.seq)
    mavlink_update_checksum(msg, msg.sysid)
    mavlink_update_checksum(msg, msg.compid)
    mavlink_update_checksum(msg, msg.msgid)

    # Thêm payload và tích lũy checksum cho payload
    for i in range(msg.len):
        byte = msg.payload[i]
        buffer.append(byte)
        mavlink_update_checksum(msg, byte)
    # Nếu có CRC extra, tích lũy thêm
    if MAVLINK_CRC_EXTRA:
        mavlink_update_checksum(msg, MAVLINK_MESSAGE_CRC_(msg.msgid))

    # Lấy checksum đã tính và ghi vào buffer (thứ tự: low byte trước, sau đó high byte)
    checksum = msg.checksum
    buffer.append(checksum & 0xFF)
    buffer.append((checksum >> 8) & 0xFF)
    offset += (6 + msg.len + 2)

    return offset